### ISSUE TITLE
Add option to "conan info" command to include build_requirements (e.g. for dependency tree)

### ISSUE THREAD

**Issue by Aalmann** [conan-io/conan#2572]
Hi @lasote and @memsharded 

it would be nice to have an extra option in the conan info command to also include the build_requirements. 
Using this flag, one could create a complete dependency graph (e.g with build_requirements in red).
Reason:
We use build_requirements instead of private requirements, because private requirements does not work as we expect. But with the current implementation it is not possible to get the full graph.
Idea:
`conan info -wbr --graph graph.html myPackage/1.2.3@user/testing`
`conan info --with-build-rrequirements --graph graph.html myPackage/1.2.3@user/testing`

Thanks.
Aalmann

**Comment by memsharded** [comment]
This is an interesting feature request.
However it might be more challenging than expected. ``conan info`` just builds the dependency graph of libraries, but it is not until later, when the real ``install`` is executed, when it is decided which things will be downloaded, which will be built from sources and which are already installed. And only at that point the ``build_requires`` are evaluated and retrieved.

We will analyze if there is some workaround that could simplify a potential implementation, without messing too much with the code. Thanks very much for the suggestion!

**Comment by Aalmann** [comment]
Would be nice to get this feature.
I checked the code but saw that it is not that easy to implement. Hope you find a way.

**Comment by olivren** [comment]
I also need this feature, that I would use for two use cases:

All our Conan recipes live in their respective Hg repo, and I wrote a script that makes it easy to clone the repo of all the dependencies of our project (it invokes `conan info` and reads the `url` attribute of the recipe to locate the repos). I would like the packages needed for build-time to be listed, too. This includes the custom Generator that we use.

This first use case is for the developers to easily have everything at hand to start modifying recipes and building new versions of a package. The second use case is to use this same list to know which packages are not used anymore, to be able to remove them from our Conan server.

**Comment by mdegel** [comment]
+1 for this feature.

We could use this as well, mainly for the traceability of the complete requirements (build_requires + requires) used in the build.

Two thoughts on how a feature like this would help:
1) Allowing to figure out how a conan package was created can be helpful when tracing errors. For example a specific build_requirement could be defective (e.g. a generator messing up build flags). If I wanted to know whether a conan package has been created by this generator, I could check the build_requirements of the conan package, instead of searching the corresponding scm repository for the version which was used to create the package (and hopefully tagged).
2) Resolving all dependencies for a conan package allows to figure out, which components need to be stored for backup. Imagine a complete flashable binary file consisting of several conan packages (OS, Libs, Applications, ...) using different types of build_requirements (different buildsystems, proprietary drivers, ...). If I wanted to store the whole dependency tree for backup (e.g. when releasing a specific version), I would have an easy way of collecting all the conan packages used.

**Comment by lasote** [comment]
@memsharded This is something I detected as a missing thing in the conaninfo.txt file when I have been investigating the "log (of a tree)" feature. Probably it is all related.

**Comment by memsharded** [comment]
Yes, most likely. We are basically talking about the reproducibility of 2 things:
- The dependency graph (including build_requires, the potential python_requires)
- The environment to reproduce the build (like env-vars)

**Comment by liberforce** [comment]
Another use case: I want to repackage some packages in conan-center we depend on for our toolchain. This is to make a bridge with our own tool, which needs in the package archive an XML file listing the dependencies and telling if they are used at runtime, buildtime, testtime. I'd like to avoid forking each recipe just for that...

I can have runtime dependencies using for example `conan info 'foo/0.1@conan/stable' --build-order=ALL`, but I miss the buildtime dependencies (and the test ones).

**Comment by monsdar** [comment]
I tested this using a simple `conanfile.txt`:

```
[requires]
zlib/1.2.11@conan/stable

[build_requires]
cmake_installer/3.13.0@conan/stable
```

Gives me the following:

```
>conan info conanfile.txt
conanfile.txt
    ID: 606fdb601e335c2001bdf31d478826b644747077
    BuildID: None
    Requires:
        zlib/1.2.11@conan/stable
    Build Requires:
        cmake_installer/3.13.0@conan/stable
cmake_installer/3.13.0@conan/stable
    ID: 456f15897172eef340fcbac8a70811f2beb26a93
    BuildID: None
    Remote: conan-center=https://api.bintray.com/conan/conan-community/conan
    URL: http://github.com/lasote/conan-cmake-installer
    License: OSI-approved BSD 3-clause
    Recipe: Cache
    Binary: Download
    Binary remote: conan-center
    Creation date: 2018-11-23 12:04:56
    Required by:
        conanfile.txt
zlib/1.2.11@conan/stable
    ID: 6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7
    BuildID: None
    Remote: conan-center=https://api.bintray.com/conan/conan-community/conan
    URL: http://github.com/conan-community/conan-zlib
    Homepage: https://zlib.net
    License: Zlib
    Author: Conan Community
    Recipe: Cache
    Binary: Download
    Binary remote: conan-center
    Creation date: 2019-03-08 09:16:33
    Required by:
        conanfile.txt
```

Do I miss something or is this feature already implemented?

**Comment by uilianries** [comment]
Hi @monsdar !

https://docs.conan.io/en/latest/devtools/build_requires.html#properties-of-build-requirements

> Build requirements are not listed in conan info nor are represented in the graph (with conan info --graph).

**Comment by monsdar** [comment]
So why is it working for me then? :)

**Comment by uilianries** [comment]
Hi @monsdar !

https://docs.conan.io/en/latest/devtools/build_requires.html#properties-of-build-requirements

    Build requirements are not listed in conan info nor are represented in the graph (with conan info --graph).

EDIT:

Because you are running for a local recipe.

**Comment by danimtb** [comment]
I think this was already implemented with the ``conan info <ref> --dry-build`` flag.

For the normal info I am getting:

```
$ conan info libuv/1.25.0@bincrafters/stable
libuv/1.25.0@bincrafters/stable
    ID: 6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7
    BuildID: None
    Remote: None
    URL: https://github.com/bincrafters/conan-libuv
    Homepage: https://github.com/libuv/libuv
    License: MIT
    Author: Bincrafters <bincrafters@gmail.com>
    Topics: conan, libuv, io, async, event
    Recipe: Cache
    Revision: 0
    Binary: Download
    Binary remote: conan-center
    Creation date: 2019-01-29 01:20:30
```

Whereas for the dry build option I see this:

```
(py37-conan) λ conan info libuv/1.25.0@bincrafters/stable -db
libuv/1.25.0@bincrafters/stable: WARN: Forced build from source
gyp_installer/20181217@bincrafters/stable: Not found in local cache, looking in remotes...
gyp_installer/20181217@bincrafters/stable: Trying with 'conan-center'...
Downloading conanmanifest.txt
[==================================================] 100B/100B
Downloading conanfile.py
[==================================================] 1.5KB/1.5KB
Downloading conan_export.tgz
[==================================================] 776B/776B
Decompressing conan_export.tgz: 100%|██████████| 776/776 [00:00<00:00, 36.1kB/s]
gyp_installer/20181217@bincrafters/stable: Downloaded recipe revision 0
gyp_installer/20181217@bincrafters/stable: WARN: Forced build from source
gyp_installer/20181217@bincrafters/stable
    ID: 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
    BuildID: None
    Remote: conan-center=https://api.bintray.com/conan/conan/conan-center
    URL: https://github.com/bincrafters/conan-gyp_installer
    Homepage: https://chromium.googlesource.com/external/gyp
    License: BSD-3-Clause
    Author: Bincrafters <bincrafters@gmail.com>
    Topics: conan, gyp, installer, meta-build-system, build-system
    Recipe: Downloaded
    Revision: 0
    Binary: Build
    Binary remote: None
    Creation date: 2018-12-17 18:28:21
    Required by:
        libuv/1.25.0@bincrafters/stable
    Build Requires:
        deployer/0.1@danimtb/testing
libuv/1.25.0@bincrafters/stable
    ID: 6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7
    BuildID: None
    Remote: None
    URL: https://github.com/bincrafters/conan-libuv
    Homepage: https://github.com/libuv/libuv
    License: MIT
    Author: Bincrafters <bincrafters@gmail.com>
    Topics: conan, libuv, io, async, event
    Recipe: Cache
    Revision: 0
    Binary: Build
    Binary remote: None
    Creation date: 2019-01-29 01:20:30
    Build Requires:
        gyp_installer/20181217@bincrafters/stable
        deployer/0.1@danimtb/testing
```

I think all the information is there and the requested features are already implemented. Could you confirm it @memsharded? Thanks!

**Comment by Aalmann** [comment]
Yes dry build does the job but sadly not in combination with build order and json e.g. to get the whole build graph of a top level dependency. One can only deep dive onto the dry build json and try to create the list.

**Comment by danimtb** [comment]
The JSON output is also available with ``--dry-build`` flag:
```
[
    {
        "revision":"0",
        "reference":"gyp_installer/20181217@bincrafters/stable",
        "is_ref":true,
        "display_name":"gyp_installer/20181217@bincrafters/stable",
        "id":"5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9",
        "build_id":null,
        "remote":{
            "name":"conan-center",
            "url":"https://api.bintray.com/conan/conan/conan-center"
        },
        "url":"https://github.com/bincrafters/conan-gyp_installer",
        "homepage":"https://chromium.googlesource.com/external/gyp",
        "license":[
            "BSD-3-Clause"
        ],
        "author":"Bincrafters <bincrafters@gmail.com>",
        "topics":[
            "conan",
            "gyp",
            "installer",
            "meta-build-system",
            "build-system"
        ],
        "recipe":"Downloaded",
        "binary":"Build",
        "creation_date":"2018-12-17 18:28:21",
        "required_by":[
            "libuv/1.25.0@bincrafters/stable"
        ]
    },
    {
        "revision":"0",
        "reference":"libuv/1.25.0@bincrafters/stable",
        "is_ref":true,
        "display_name":"libuv/1.25.0@bincrafters/stable",
        "id":"6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7",
        "build_id":null,
        "remote":{
            "name":"conan-center",
            "url":"https://api.bintray.com/conan/conan/conan-center"
        },
        "url":"https://github.com/bincrafters/conan-libuv",
        "homepage":"https://github.com/libuv/libuv",
        "license":[
            "MIT"
        ],
        "author":"Bincrafters <bincrafters@gmail.com>",
        "topics":[
            "conan",
            "libuv",
            "io",
            "async",
            "event"
        ],
        "recipe":"Downloaded",
        "binary":"Build",
        "creation_date":"2019-01-29 01:20:30",
        "build_requires":[
            "gyp_installer/20181217@bincrafters/stable"
        ]
    }
]
```
The ``--build-order`` flag results can be also stored in a JSON file:

``conan info libuv/1.25.0@bincrafters/stable --json info.json --build-order libuv/1.25.0@bincrafters/stable``
```
{
    "groups":[
        [
            "libuv/1.25.0@bincrafters/stable"
        ]
    ]
}
```

And finally, I haven't checked with a larger graph, but I think it is also working in combination:

```
$ conan info libuv/1.25.0@bincrafters/stable --json info.json --build-order libuv/1.25.0@bincrafters/stable -db

$ cat info.json
{"groups": [["libuv/1.25.0@bincrafters/stable"]]}
```

But I am not sure if the build order will be helpful in combination with the dry-build command, as build requirements don't affect the package ID (by default), meaning that downstream packages wouldn't need to be rebuilt.

Please, let us know if your use case is covered with the above. Thanks!

**Comment by Aalmann** [comment]
@danimtb 
As described above I'm aware of the `--dry-build` in combination with `--json`, but with a larger graph you will not get the build order for the build_requires packages, because they were only present at the package level and I have to loop the nested dicts, collecting everything together etc.. Let's say there are more build requirements and all of them needs to be rebuilt ( e.g. for a new platform / new architecture). Then it is required to get the build order of the whole dependency graph including the build_requirements which (may) be recreated too.
So what I would like to have is the ability to get the whole graph (with --dry-build) as build order (--build-order xyz).

**Comment by memsharded** [comment]
Conan 2.0 now always compute the full graph, including build-requires, and those are included in build-order too.
